In [20]:
import json
from pprint import pprint

In [21]:
with open('recipes.json') as f:
    recipes = json.load(f)

In [49]:
%%time
for i in range(1000):
    cuisines = []  # создаём пустой список для хранения уникальных значений кухонь
    for recipe in recipes:  # начинаем перебор всех рецептов
       if not(recipe['cuisine'] in cuisines):  # если тип кухни текущего блюда ещё не встречался
            cuisines.append(recipe['cuisine']) # добавляем его к списку cuisines
len(cuisines)

CPU times: user 99.2 ms, sys: 0 ns, total: 99.2 ms
Wall time: 98 ms


20

In [50]:
%%time
for i in range(1000):
    cuisines = set()
    for recipe in recipes:
        cuisines.add(recipe['cuisine'])
    len(cuisines)

CPU times: user 83.2 ms, sys: 0 ns, total: 83.2 ms
Wall time: 82 ms


#### Task 1.
Сколько ингредиентов включено в состав всех блюд, описанных в наборе данных?

Подсказка: Обратите внимание, что для добавления в список или множество нового ингредиента вам нужно будет организовать перебор ингредиентов каждого блюда в отдельном вложенном цикле.

#### Solution:

In [54]:
ingredients = set()
for recipe in recipes:
    ingredients = ingredients | set(recipe['ingredients'])
len(ingredients)

1318

#### Conclusion: 1318 is the right answer!!!

#### Task 2.
Сколько ингредиентов встречается в рецептах блюд итальянской кухни ('cuisine' = italian)?

#### Solution:

In [55]:
ingredients = set()
for recipe in recipes:
    if recipe['cuisine'] == 'italian':
        ingredients = ingredients | set(recipe['ingredients'])
len(ingredients)

406

#### Conclusion: 406 is the right answer!!!

#### Task 3.
Какие ингредиенты не встречаются в рецептах блюд русской кухни ('cuisine' = 'russian')?
- bacon
- bread slices
- buttermilk
- red beets
- mozzarella cheese
- carrots

#### Solution:

In [56]:
ingredientns_to_find = {
    'bacon',
    'bread slices',
    'buttermilk',
    'red beets',
    'mozzarella cheese',
    'carrots',
}

russian_ingredients = set()
for recipe in recipes:
    if recipe['cuisine'] == 'russian':
        russian_ingredients = russian_ingredients | set(recipe['ingredients'])

ingredientns_to_find.difference(russian_ingredients)

{'bacon', 'bread slices', 'carrots'}

#### Conclusion: {'bacon', 'bread slices', 'carrots'} is the right answer!!!

In [59]:
ingredients = set()

for recipe in recipes:
    ingredients = ingredients | set(recipe['ingredients'])

food = dict()
for ingredient in ingredients:
    food[ingredient] = 0

for recipe in recipes:
    for ingredient in recipe['ingredients']:
        food[ingredient] += 1

#### Task 4.
Какие ингредиенты входят в состав более чем 20% блюд, представленных в наборе данных?

Подсказка: Напоминаем, что весь набор данных содержит сведения о 500 блюдах.
- water
- olive oil
- salt
- garlic
- pepper 

#### Solution:

In [63]:
ingredients_to_check = {
    'water',
    'olive oil',
    'salt',
    'garlic',
    'pepper'
}

more_20pct_ingredients = set()

for ingredient in ingredients_to_check:
    if food[ingredient] > 100:
        more_20pct_ingredients.add(ingredient)

more_20pct_ingredients

{'garlic', 'salt'}

#### Conclusion: {'garlic', 'salt'} is the right answer!!!

#### Task 5.
Какой ингредиент входит в состав самого большого количества блюд?

#### Solution:

In [64]:
max_count = 0
most_often_used_ingredient = ''

for ingredient in ingredients:
    if food[ingredient] > max_count:
        max_count = food[ingredient]
        most_often_used_ingredient = ingredient

most_often_used_ingredient

'salt'

#### Conclusion: 'salt' is the right answer!!!

#### Task 6.
Сколько ингредиентов входит в состав только одного блюда?

#### Solution:

In [66]:
one_food_count = 0
for ingredient in ingredients:
    if food[ingredient] == 1:
        one_food_count += 1
one_food_count

684

#### Conclusion: 684 is the right answer!!!